_posts/2018-10-09-enabling-ecc-on-pc-engines-apu2.md

```diff
@@ -0,0 +1,354 @@
+this register's value in coreboot as well. After doing so ECC errors were
+reported:
+
+```
+(...)
+2018-10-01 01:01:16 - Running test #3 (Test 3 [Moving inversions, ones & zeroes])
+2018-10-01 01:01:16 - MtSupportRunAllTests - Injecting ECC error
+2018-10-01 01:01:16 - inject_amd64 - new nb_arr_add = 80000000
+2018-10-01 01:01:16 - inject_amd64 - new dram_ecc = 0012000F
+2018-10-01 01:01:16 - MCA NB Status High=00000000
+2018-10-01 01:01:16 - inject_amd64 - new nb_arr_add = 80000002
+2018-10-01 01:01:16 - inject_amd64 - new dram_ecc = 0012000F
+2018-10-01 01:01:16 - MCA NB Status High=00000000
+2018-10-01 01:01:16 - inject_amd64 - new nb_arr_add = 80000004
+2018-10-01 01:01:16 - inject_amd64 - new dram_ecc = 0012000F
+2018-10-01 01:01:16 - MCA NB Status High=846FC000
+2018-10-01 01:01:16 - MtSupportRunAllTests - Setting random seed to 0x50415353
+2018-10-01 01:01:16 - MtSupportRunAllTests - Start time: 515 ms
+2018-10-01 01:01:16 - ReadMemoryRanges - Available Pages = 1035000
+2018-10-01 01:01:16 - MtSupportRunAllTests - Enabling memory cache for test
+2018-10-01 01:01:16 - MtSupportRunAllTests - Enabling memory cache complete
+2018-10-01 01:01:16 - Start memory range test (0x0 - 0x12F000000)
+2018-10-01 01:01:16 - Pre-allocating memory ranges >=16MB first...
+2018-10-01 01:01:16 - All memory ranges successfully locked
+2018-10-01 01:01:17 - MCA NB Status=846FC000F2080813 
+2018-10-01 01:01:17 - MCA NB Address=00000000CFE528D0
+2018-10-01 01:01:17 - [MEM ERROR - ECC] Test: 3, Address: CFE528D0, ECC Corrected: yes, Syndrome: F2DF, Channel/Slot: N/A
+2018-10-01 01:02:21 - MCA NB Status=846FC000F2080813 
+2018-10-01 01:02:21 - MCA NB Address=00000000CE320540
+2018-10-01 01:02:21 - [MEM ERROR - ECC] Test: 3, Address: CE320540, ECC Corrected: yes, Syndrome: F2DF, Channel/Slot: N/A
+2018-10-01 01:02:36 - Cleanup - Releasing all memory ranges...
+2018-10-01 01:02:36 - MtSupportRunAllTests - Test execution time: 80.088 (Test 3 cumulative error count: 0)
+(...)
+```
+
+## Summary and additional findings
+
+After the last fix, ECC is enabled as well as ECC error injection on all
+supported hardware (that is, every APU platform with 4 GB of memory). Generated
+reports should look like this:
+
+![herebescreenshot]
+
+Every corrected ECC error has the same syndrome - F2DF. It is caused by
+MemTest86 setting D18F3xBC_x8 (DRAM ECC) to `0012000F`. More info about the
+meaning of these is available in [BKDG](https://www.amd.com/system/files/TechDocs/52740_16h_Models_30h-3Fh_BKDG.pdf)
+on pages 172-174 (ECC syndromes) and 456 (DRAM ECC register, NB Array Address).
+
+Another thing is that sometimes more than one ECC error for a test is injected.
+It is caused by internal work of injection - only bits inside a cache line can
+be chosen, but an error is injected on next non-cached operation at accessed
+address, which can change or not between any of three attempts to inject
+an error.
```

Comment:
@krystian-hebel great reading! I would add at the end future plans, where is code that we implemented and that we will push it upstream soon.

Reply:
done